Update controller runtime to 0.19.3

ChunyiLyu opened the pull request:
# Context

Closes: #252 

This PR updates controller runtime from `0.18.4` to `0.19.3`. It includes changes:

### Migration to new Webhook
* migrating off deprecated `webhook.Validator` and `webhook.Default` to CustomValidator and CustomDefaulter for promises and promise releases.
* move webhooks to `internal/webhook`.  Since kubebuilder `4.3.0`, they've also changed the generated project structure that webhook now resides in `internal/webhook` rather than `api/`. 

Migration steps are documented in kubebuilder release notes: https://github.com/kubernetes-sigs/kubebuilder/releases

### Compatibility changes

* upgrade to latest `controller-gen` to resolve errors
```
❯ make manifests
...k8s.io/api@v0.31.0/core/v1/doc.go:21:1: missing argument "" (at <input>)=go/pkg/mod/k8s.io/api@v0.31.0/core/v1/doc.go:21:1: missing argument "" (at <input>)
k8s.io/api/core/v1:-: unknown type "k8s.io/api/core/v1".SecretReference
github.com/syntasso/kratix/api/v1alpha1:-: unable to locate schema for type "k8s.io/api/core/v1".SecretReference
Error: not all generators ran successfully
run `controller-gen crd:ignoreUnexportedFields=true rbac:roleName=manager-role webhook paths=./... output:crd:artifacts:config=config/crd/bases -w` to see all available markers, or `controller-gen crd:ignoreUnexportedFields=true rbac:roleName=manager-role webhook paths=./... output:crd:artifacts:config=config/crd/bases -h` for usage
make: *** [manifests] Error 1
```

* turn off controller name validation in `controllers/` tests to resolve
```
 [FAILED] Unexpected error:
      <*errors.errorString | 0x140007061c0>:
      controller with name redis already exists. Controller names must be unique to avoid multiple controllers reporting to the same metric
      {
          s: "controller with name redis already exists. Controller names must be unique to avoid multiple controllers reporting to the same metric",
      }
  occurred
  In [It] at: /kratix/controllers/promise_controller_test.go:484 @ 12/06/24 14:08:08.644
```
This is a new validation added to controller runtime 0.19 to make ensure no controller is registered with the same name in a manager. You can turn it off at the manager level for all controllers registered with the manager. We do not need it in controller tests as we may call `reconcileTilCompletion` within the same test spec.

## Follow up stories
1. Repeat the same in other operator we have.
2. https://github.com/syntasso/kratix/issues/293
3. https://github.com/syntasso/kratix/issues/295

kirederik commented:
I'm assuming these changes have no impact on the distribution.yaml that gets generated right? If not, LGTM!

ChunyiLyu commented:
@kirederik diff being 

1. updated controller-gen annotations
2. updated description for status.conditions
3. aggregated roles resources
```
  resources:
  - destinations/status
  - promisereleases/status
  - promises/status
  - workplacements/status
  ```
These are used to be in their own section. No functional difference.
